src/db/pool.cr

```diff
@@ -2,47 +2,88 @@ require "weak_ref"
 
 module DB
   class Pool(T)
+    # Pool configuration
+
+    # initial number of connections in the pool
     @initial_pool_size : Int32
-    # maximum amount of objects in the pool. Either available or in use.
+    # maximum amount of connections in the pool (Idle + InUse)
     @max_pool_size : Int32
-    @available = Set(T).new
-    @total = [] of T
+    # maximum amount of idle connections in the pool
+    @max_idle_pool_size : Int32
+    # seconds to wait before timeout while doing a checkout
     @checkout_timeout : Float64
-    # maximum amount of retry attempts to reconnect to the db. See `Pool#retry`.
+    # maximum amount of retry attempts to reconnect to the db. See `Pool#retry`
     @retry_attempts : Int32
+    # seconds to wait before a retry attempt
     @retry_delay : Float64
 
+    # Pool state
+
+    # total of open connections managed by this pool
+    @total = [] of T
+    # connections available for checkout
+    @idle = Set(T).new
+    # connections waiting to be stablished (they are not in *@idle* nor in *@total*)
+    @inflight : Int32
+
+    # Sync state
+
+    # communicate that a connection is available for checkout
+    @availability_channel : Channel(Nil)
+    # signal how many existing connections are waited for
+    @waiting_resource : Int32
+    # global pool mutex
+    @mutex : Mutex
+
     def initialize(@initial_pool_size = 1, @max_pool_size = 0, @max_idle_pool_size = 1, @checkout_timeout = 5.0,
                    @retry_attempts = 1, @retry_delay = 0.2, &@factory : -> T)
-      @initial_pool_size.times { build_resource }
-
       @availability_channel = Channel(Nil).new
       @waiting_resource = 0
+      @inflight = 0
       @mutex = Mutex.new
+
+      @initial_pool_size.times { build_resource }
+    end
+
+    def stats
+      {open_connections: @total.size}
     end
 
     # close all resources in the pool
     def close : Nil
       @total.each &.close
       @total.clear
-      @available.clear
+      @idle.clear
+    end
+
+    # Returns stats of the pool
+    def stats
+      {open_connections: @total.size}
```

Comment:
Are there any other stats that should be returned?
If not should there just be a method called `open_connections`?

Reply:
I expect to add some but in a future PR. I needed it to write the spec in crystal-mysql

Reply:
what else would be added?

Reply:
https://github.com/golang/go/blob/master/src/database/sql/sql.go#L990-L1026 🙈

Reply:
This should better be a struct.

Reply:
isnt a named tuple a fancy read only struct?

Reply:
Yes. They are a struct. The GC.stats in the std-lib returns a record/struct. The difference is how to access the information. `s.prop` vs `s[:prop]`, how one would use `responds_to?`/`has_key?` and the ability to merge or shrink information in the later.

@straight-shoota is there any arguments why it should be better a struct? I like named tuples   over records usually for giving raw information.

Reply:
If nothing else, a struct can be easily documented what each property means. And I agree with what IIRC @asterite said, that the main use case for named tuples is named arguments and for every thing else it's probably better to use a defined data type.

Reply:
There, the stats are now a `record`. Rebased and ready to merge upon 👍